posts/history_download/index.md

```diff
@@ -0,0 +1,316 @@
+---
+title: "Panel AI Chatbot Tips: Memory and Downloadable Conversations"
+date: "2023-12-22"
+description: ""
+author:
+  - Andrew Huang
+  - Sophia Yang
+categories: [showcase, panel, ai, llm, chatbot]
+image: "images/chatbot.png"
+---
+
+In this blog post, we'll explore how to build a simple AI chatbot, enhance it with memory capabilities, and finally, implement a feature to download conversations for further fine-tuning.
+
+We will cover: 
+- Getting started building a simple AI chatbot (with no memory)
+- Adding memory to manage chat histories 
+- Adding download button to download all conversations 
+
+
+Before we get started, let's first make sure we install needed packages like `panel`, `mistralai`, `openai` in our Python environment and save our API keys as environment variables: 
+
+`export MISTRAL_API_KEY="TYPE YOUR API KEY"`
+
+`export OPENAI_API_KEY="TYPE YOUR API KEY"`
+
+# Getting started building a simple AI chatbot (with no memory)
+
+## Mistral models
+
+In this blog post, we will only use the Mistral API. If you are interested in using Mistral models locally, check out our previous blog post [Build a Mixtral Chatbot with Panel](https://blog.holoviz.org/posts/mixtral/) to see how we used Mistral API, transformers, llama.cpp, and Panel to create AI chatbots that use the Mixtral 8x7B Instruct model. 
+
+When we do not need to keep our conversation history, we are only sending one round of user message to model. Thus, in this example, the `messages` that get sent to the model are defined as `[ChatMessage(role="user", content=contents)]`. 
+
+```python
+import os
+import panel as pn
+from mistralai.client import MistralClient
+from mistralai.models.chat_completion import ChatMessage
+
+pn.extension()
+
+
+async def callback(contents: str, user: str, instance: pn.chat.ChatInterface):
+    model = "mistral-small"
+    messages = [
+        ChatMessage(role="user", content=contents)
+    ]
+    print(messages)
+    response = client.chat_stream(model=model, messages=messages)
+
+    message = ""
+    for chunk in response:
+        part = chunk.choices[0].delta.content
+        if part is not None:
+            message += part
+            yield message
+
+
+client = MistralClient(api_key=os.environ["MISTRAL_API_KEY"])
+chat_interface = pn.chat.ChatInterface(callback=callback, callback_user="Mixtral")
+chat_interface.send(
+    "Send a message to get a reply from Mixtral!", user="System", respond=False
+)
+chat_interface.servable()
+```
+
+<img src="./images/mistral.png" width="100%" style="margin-left: auto; margin-right: auto; display: block;"></img>
+
+## OpenAI models
+The code of using OpenAI models looks very similar. We are using OpenAI's API with async/await to use the asynchronous client. To use async, we simply import `AsyncOprnAI` instead of `OpenAI` and add `await` with the API call. 
```

Comment:
```suggestion
The code of using OpenAI models looks very similar. We are using OpenAI's API with async/await to use the asynchronous client. To use async, we simply import `AsyncOpenAI` instead of `OpenAI` and add `await` with the API call. 
```